### ISSUE TITLE
tf.contrib.eager.Saver.__init__ var_list argument should be optional

### ISSUE THREAD

**Issue by traviskaufman** [tensorflow/tensorflow#24234]
**System information**
- TensorFlow version (you are using): 1.12.0
- Are you willing to contribute it (Yes/No): Yes!

**Describe the feature and the current behavior/state.**

[tf.contrib.eager.Saver](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/Saver) (aka `tfe.Saver`) claims to be "A tf.train.Saver adapter for use when eager execution is enabled." However, the current api for `__init__()` requires that a `var_list` argument be passed in. 

Looking at [the code](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/contrib/eager/python/saver.py#L138), the `var_list` argument is simply delegated to `tf.train.Saver`. The [tf.train.Saver](https://www.tensorflow.org/api_docs/python/tf/train/Saver#__init__) does _not_ require `var_list` to be pass in, having it default to `None`. I propose that `tfe.Saver` do the same.

**Will this change the current api? How?**

Yes, but it is a backwards-compatible change. Users of `tfe.Saver` will no longer have to specify a `var_list` argument. Any arguments already present will not change as they will override the default value.

**Who will benefit with this feature?**

My use-case for this feature is that I've written a custom estimator that's run on Cloud ML Engine, and I would like to write a local script that can restore the estimator and be evaluated against held out set of data different from the evaluation set the estimator was trained/tuned with. I'd like to take advantage of eager execution to do this. Not having to invoke `tfe.Saver` with a spurious `None` argument would make the code align with the original `tf.train.Saver` APIs.

**Any Other info.**

Thanks for the consideration! I know the team gets inundated with a ton of issues and as I mentioned above, I'm more than happy to make the change.

It also looks like there's a lot of other kwargs missing from `tfe.Saver`...perhaps it should just accept `*args` and `**kwargs` and forward them to `tf.train.Saver`?

**Comment by allenlavoie** [comment]
The issue is that we've removed collections when eager execution is enabled, including the variables collection. Generally collections are just memory leaks waiting to happen executing eagerly, and the global behavior can be error-prone.

Could you take a look at [object-based saving](https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint) and see if it works for you? That will collect variables through object dependencies without relying on global collections. It has been designed to be pleasant to use when executing eagerly, for example restoring variables on creation. We're moving away from `tf.train.Saver`, and `tfe.Saver` will be deprecated in the next release.

**Comment by traviskaufman** [comment]
> Generally collections are just memory leaks waiting to happen executing eagerly, and the global behavior can be error-prone.

Makes sense. Thanks for pointing me to the `Checkpoint` reference! I actually think I don't even need that, since my model is an `Estimator` rather than `tf.keras.Model`. I realized I can just pass `tf.train.latest_checkpoint(...)` to `warm_start_from` and get it working:

```python
estimator = tf.estimator.Estimator(
  model_fn,
  warm_start_from=tf.train.latest_checkpoint(checkpoint_dir))
```

**Comment by allenlavoie** [comment]
We have [a guide which includes collecting variables](https://www.tensorflow.org/alpha/guide/checkpoints) now. tfe.* is going away, and Estimator works fine as you say. So I'm closing this as fixed/obsolete.
